File: 2022_11_03_leetcode_in_fsharp/LeetCode/P0001_TwoSum.fs
// url: https://leetcode.com/problems/two-sum/
// tags: collect, find
// examples: [| 2; 7; 11; 15 |] -> 9

module LeetCode.P0001_TwoSum

open Utils

let twoSum nums target =
    { 0 .. Array.length nums - 1 }
    |> Seq.collect (fun i -> { 0 .. i - 1 } |> Seq.map (fun j -> (j, i)))
    |> Seq.find (fun (j, i) -> nums.[i] + nums.[j] = target)

twoSum [| 2; 7; 11; 15 |] 9 === (0, 1)
twoSum [| 3; 2; 4 |] 6 === (1, 2)
twoSum [| 3; 3 |] 6 === (0, 1)

File: 2022_11_19_leetcode_in_csharp/LeetCode/P0001_TwoSum.cs
// https://leetcode.com/problems/two-sum/

namespace LeetCode;

using static System.Linq.Enumerable;

class P0001_TwoSum
{
    static int[] TwoSum(int[] nums, int target)
    {
        return Range(0, nums.Length)
            .SelectMany(i => Range(0, i), (i, j) => new[] { j, i })
            .First(res => nums[res[0]] + nums[res[1]] == target)
            .ToArray();
    }

    public static void Run()
    {
        var inputs = new[]
        {
            (new []  {2, 7, 11, 15}, 9),
            (new []  {3, 2, 4}, 6),
            (new []  {3, 3}, 6)
        };

        foreach (var (nums, target) in inputs)
        {
            var output = TwoSum(nums, target);
            Console.WriteLine($" {nums.JoinToString()} - {target} -> {output.JoinToString()}");
        }
    }
}
